compiler/resolution/nilChecking.cpp

```diff
@@ -0,0 +1,782 @@
+    lhsSym = initSe->symbol();
+    rhsExpr = call->get(2);
+    moveLike = true;
+    userCall = toCallExpr(rhsExpr);
+  } else if (call->isNamedAstr(astrSequals)) {
+    initSe = toSymExpr(call->get(1));
+    lhsSym = initSe->symbol();
+    rhsExpr = call->get(2);
+    moveLike = true;
+    assign = true;
+    userCall = call;
+  } else if (isRecordInitOrReturn(call, initSe, initCall)) {
+    lhsSym = initSe->symbol();
+    rhsExpr = initCall;
+    moveLike = true;
+    userCall = call;
+  } else {
+    userCall = call;
+  }
+
+  FnSymbol* userCalledFn = NULL;
+  if (userCall) {
+    userCalledFn = userCall->resolvedOrVirtualFunction();
+  }
+
+  bool ignoreGlobalUpdates = false;
+
+  // raise errors if the call is problematic somehow
+  if (userCall && raiseErrors) {
+    checkForNilDereferencesInCall(userCall, OUT);
+  }
+
+  if (moveLike) {
+    // lhsSym and rhsExpr set above
+
+    if (lhsSym->isRef()) {
+      // move-ish, but lhs is a ref
+
+      if (assign && isClassIshType(lhsSym)) {
+        // e.g. owned = nil
+        SymExpr* se = toSymExpr(call->get(2));
+        Symbol* rhsSym = se->symbol();
+        Symbol* referent = getReferent(lhsSym, OUT);
+        if (referent)
+          OUT[referent] = aliasLocationCopyValueFrom(rhsSym, OUT);
+        else if (lhsSym->hasFlag(FLAG_RETARG) == false)
+          storeToUnknown = true;
+
+      } else if (userCalledFn != NULL) {
+        // e.g.
+        //   ref r = returnsRefArg(x);
+        // We can't say what 'r' refers to without analyzing that fn
+        //   - it could refer to a global
+        //   - it could refer to a by-ref argument
+        //   - it could refer to a field in a heap-allocated class
+
+        // This could be another class, ref-to-unknown, if it helps
+        OUT[lhsSym] = unknownAliasLocation();
+      } else {
+        if (userCall == NULL) {
+          // Acts like a move
+          SymExpr* se = toSymExpr(call->get(2));
+          Symbol* rhsSym = se->symbol();
+          if (call->isPrimitive(PRIM_MOVE)) {
+            OUT[lhsSym] = OUT[rhsSym];
+          } else {
+            Symbol* referent = getReferent(lhsSym, OUT);
+            if (referent)
+              OUT[referent] = aliasLocationCopyValueFrom(rhsSym, OUT);
+            else if (lhsSym->hasFlag(FLAG_RETARG) == false)
+              storeToUnknown = true;
+          }
+        } else {
+          if (userCall->isPrimitive(PRIM_ADDR_OF) ||
+              userCall->isPrimitive(PRIM_SET_REFERENCE)) {
+            SymExpr* se = toSymExpr(userCall->get(1));
+            Symbol* rhs = se->symbol();
+            if (rhs->isRef())
+              // if rhs is a ref, just propagate any MUST_ALIAS_REFVAR
+              OUT[lhsSym] = OUT[rhs];
+            else
+              // otherwise, make a new MUST_ALIAS_REFVAR
+              OUT[lhsSym] = refAliasLocation(se->symbol());
+          } else {
+            OUT[lhsSym] = unknownAliasLocation();
+          }
+        }
+      }
```

Comment:
Let's label the above close-curly with the condition that it closes. It looked to me it was `lhsSym->isRef()` , however judging by the "move-ish" comment below that's not the case. So, clarification is needed!